esx_service/vmdk_ops.py

```diff
@@ -777,31 +776,90 @@ def disk_attach(vmdk_path, vm):
+
+            if len(availSlots) == 0:
+                idx = idx+1
+            else:     
+                disk_slot = availSlots.pop()
+                pci_slot_number = get_controller_pci_slot(vm, pvsci[idx],
+                                                          offset_from_bus_number)
+                 
+                logging.debug("Find an available slot: controller_key = %d slot = %d", controller_key, disk_slot)
+
+        # search through all the existing pvscsi controller, and cannot find a slot
+        # try to add another pvscsi controller if possible        
+        if ((disk_slot is None) and (len(availSlots) == 0)):
+            disk_slot = 0  # starting on a fresh controller
+            if len(controllers) >= max_scsi_controllers:
+                msg = "Failed to place new disk - out of disk slots"
+                logging.error(msg + " VM=%s", vm.config.uuid)
+                return err(msg)
+                    
+            # find empty bus slot for the controller:
+            taken = set([c.busNumber for c in controllers])
+            avail = set(range(0, max_scsi_controllers)) - taken
+
+            key = avail.pop()  # bus slot
+            controller_key = key + offset_from_bus_number
+            controller_spec = vim.VirtualDeviceConfigSpec(
+            operation='add',
+            device=vim.ParaVirtualSCSIController(key=controller_key,
+                                                 busNumber=key,
+                                                 sharedBus='noSharing', ), )
+            # changes spec content goes here
+            pvscsi_change = []
+            pvscsi_change.append(controller_spec)
+            spec = vim.vm.ConfigSpec()
+            spec.deviceChange = pvscsi_change
+
+            try:
+                wait_for_tasks(si, [vm.ReconfigVM_Task(spec=spec)])
+            except vim.fault.VimFault as ex:
+                msg=("Failed to add PVSCSI Controller: %s", ex.msg)
+                return err(msg)
+            
+            logging.info("Add a new pvscsi controller, id=%d", controller_key) 
+
+            # get all the pvscsi controller
+            devices = vm.config.hardware.device
+            pvsci = [d for d in devices
+              if type(d) == vim.ParaVirtualSCSIController and
+                 d.key == controller_key]
 
+            new_idx = None  
+            for x in range (0, len(pvsci)):
+                if (pvsci[x].key == controller_key):
+                    new_idx = x
+                    break
+                     
+            # get the pci_slot_number of newly added pvscsi controller  
+            logging.debug("idx for newly added pvscsi controller is %d", new_idx)
+            pci_slot_number = get_controller_pci_slot(vm, pvsci[new_idx],
+                                                          offset_from_bus_number)
+ 
+  
     # add disk as independent, so it won't be snapshotted with the Docker VM
```

Comment:
Overall, please refactor the code into new functions so that this whole logic is easy to understand and have logic to add a new controller (whether or not the VM has pvscsi controllers) only if are no free disk slots and we are within limits of the max number of scsi controllers.